cmd/frontend/internal/httpapi/graphql.go

```diff
@@ -49,16 +50,37 @@ func actorTypeLabel(isInternal, anonymous bool, requestSource trace.SourceType)
 	return "unknown"
 }
 
-func exceedsLimit(costValue, limitValue int) bool {
-	return costValue > limitValue
+type violationInfo struct {
+	violationType string
+	actual        int
+	limit         int
 }
 
-func writeViolationError(w http.ResponseWriter) error {
+func exceedsLimit(costValue, limitValue int, violationType string) (bool, *violationInfo) {
+	if costValue > limitValue {
+		return true, &violationInfo{
+			violationType: violationType,
+			actual:        costValue,
+			limit:         limitValue,
+		}
+	}
+
+	return false, nil
+}
+
+func writeViolationError(w http.ResponseWriter, info violationInfo) error {
 	w.WriteHeader(http.StatusBadRequest) // 400 because retrying won't help
 	return writeJSON(w, graphql.Response{
 		Errors: []*gqlerrors.QueryError{
 			{
-				Message: "query exceeds maximum query cost",
+				Message: fmt.Sprintf("Query exceeds maximum %s limit", info.violationType),
+				Extensions: map[string]interface{}{
+					"code":     "errQueryComplexityLimitExceeded",
+					"type":     info.violationType,
+					"limit":    info.limit,
+					"actual":   info.actual,
+					"docs_url": "https://sourcegraph.com/docs/api/graphql#cost-limits",
```

Comment:
if we do something like ```suggestion
					"docs_url": conf.Get().ExternalURL + "/help/api/graphql#cost-limits",
```
i think this would redirect to the right version of the docs page for the running version

Reply:
that works! thanks :+1: